### PR TITLE
[VM] Check for Gas Meter Safety

### PR DESCRIPTION
**Pull request by darioncassel**
## Motivation

Add checks for safety of `calculate_transaction_gas` in `gas_meter.rs`.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

Testing via `cargo test`. Passes existing tests. To reproduce the analysis performed by MIRAI, please run `RUSTC_WRAPPER=mirai MIRAI_LOG=warn cargo build --features "mirai-contracts"` from the VM directory.

## Related PRs

N/A

### CODE REVIEW

File: language/vm/vm_runtime/src/foreign_contracts.rs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This file contains models of the vm crate's dependencies for use with MIRAI.
+
+pub mod types {
+    pub mod transaction {
+        pub const MAX_TRANSACTION_SIZE_IN_BYTES: usize = 4096;
```
Comment (hermanventer) [inline language/vm/vm_runtime/src/foreign_contracts.rs]: Why not just define this as a u64?

Comment (darioncassel) [reply language/vm/vm_runtime/src/foreign_contracts.rs]: This is a foreign contract for that constant.

File: language/vm/vm_runtime/src/lib.rs [hunk 2]
```diff
@@ -123,6 +123,8 @@ pub mod txn_executor;
 pub mod execution_stack;
 #[cfg(not(feature = "instruction_synthesis"))]
 mod execution_stack;
+#[macro_use]
+extern crate mirai_annotations;
```
Comment (tzakian) [inline language/vm/vm_runtime/src/lib.rs]: Probably a good reason, but Is this extern crate needed?

Comment (hermanventer) [reply language/vm/vm_runtime/src/lib.rs]: Yes, this crate defines annotations such as precondition!. It has no impact on the runtime of release builds.

**Pull request merged**